traversal/procedure/ProcedureVertex.java

```diff
@@ -175,124 +189,347 @@ <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> iterateAndFilter
         ) {
-            // TODO we should be using forward() to optimise filtering for >, <, and =
-            assert id().isVariable();
-            for (Predicate.Value<?, ?> predicate : props().predicates()) {
-                if (Objects.equals(predicate, exclude)) continue;
-                for (Traversal.Parameters.Value value : parameters.getValues(id().asVariable(), predicate)) {
-                    iterator = iterator.filter(a -> predicate.apply(a.asAttribute(), value));
+            if (props().predicates().isEmpty()) return iterateSorted(edges, order);
+            else {
+                if (iterate(edges).anyMatch(kv -> kv.key().isAttribute() && kv.key().asAttribute().type().valueType().equals(STRING))) {
+                    return iterateSorted(filterPredicatesAndMapEdges(edges, params, ThingVertex::asAttribute), order);
+                } else {
+                    return iterateSorted(filterPredicatesAndMapEdges(edges, params, v -> v.asAttribute().toValue()), order);
                 }
             }
-            return iterator;
         }
 
-        Forwardable<KeyValue<ThingVertex, ThingVertex>, Order.Asc> filterPredicatesOnEdge(
-                Forwardable<KeyValue<ThingVertex, ThingVertex>, Order.Asc> iterator, Traversal.Parameters parameters
+        private TreeSet<KeyValue<ThingVertex, ThingVertex>> filterPredicatesAndMapEdges(
+                List<KeyValue<ThingVertex, ThingVertex>> edges,
+                Traversal.Parameters params, Function<ThingVertex, AttributeVertex<?>> mapper
+        ) {
+            TreeSet<KeyValue<ThingVertex, ThingVertex>> filtered = new TreeSet<>();
+            edges.forEach(kv -> {
+                if (checkPredicates(kv.key(), params, set())) {
+                    filtered.add(new KeyValue<>(mapper.apply(kv.key()), kv.value()));
+                }
+            });
+            return filtered;
+        }
+
+        <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> mergeAndFilterPredicatesOnVertices(
+                GraphManager graphMgr, List<Pair<TypeVertex, Forwardable<ThingVertex, ORDER>>> vertexIters,
+                Traversal.Parameters params, ORDER order
+        ) {
+            return mergeAndFilterPredicatesOnVertices(graphMgr, vertexIters, params, order, false);
+        }
+
+        <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> mergeAndFilterPredicatesOnVertices(
+                GraphManager graphMgr, List<Pair<TypeVertex, Forwardable<ThingVertex, ORDER>>> vertexIters,
+                Traversal.Parameters params, ORDER order, boolean forceValueSort
+        ) {
+            if (props().predicates().isEmpty() && !forceValueSort) {
+                return merge(iterate(vertexIters).map(Pair::second), order);
+            } else {
+                if (iterate(vertexIters).anyMatch(pair -> pair.first().isAttributeType() && pair.first().asType().valueType().equals(STRING))) {
```

Comment:
we have to exclude `String` attributes from using optimised predicates, they just have to use filtering based on predicates. This is because `String` attributes do not have the right native value-sortedness yet!

Value-sorted vertices can use optimised predicates which leverage `forward()` to skip ranges of vertices that are less than the required value, and stop early when the value is hit without retrieving any more.